File: L1S/v3/code/ocaml/src/ir_build.ml
type builder = {
  source: Span.source_file;
  diags: Diagnostic.t list;
}

let make source = { source; diags = [] }

let ast_type_to_ir = function
  | Ast.Prim Ast.I32 -> Ir.I32 | Ast.Prim Ast.I64 -> Ir.I64
  | Ast.Prim Ast.U32 -> Ir.U32 | Ast.Prim Ast.U64 -> Ir.U64
  | Ast.Prim Ast.I8 -> Ir.I8 | Ast.Prim Ast.U8 -> Ir.U8
  | Ast.Prim Ast.Bool -> Ir.Bool | Ast.Prim Ast.Void -> Ir.Void
  | Ast.Ptr _ -> Ir.Ptr Ir.I32
  | Ast.Named _ -> Ir.I32
  | Ast.Array _ -> Ir.Ptr Ir.I32

let build_func func =
  let ret = ast_type_to_ir func.Ast.return_type in
  let ir_func = Ir.make_function func.Ast.name ret in
  ir_func.Ir.params <- List.map (fun p -> ast_type_to_ir p.Ast.type_) func.Ast.params;
  let entry = Ir.make_block "b1" in
  let vars = Hashtbl.create 16 in
  (* Allocate stack slots for parameters *)
  List.iter (fun p ->
    let v = Ir.fresh_vreg ir_func in
    Ir.(entry.instrs <- Alloc (v, ast_type_to_ir p.Ast.type_) :: entry.instrs);
    let pt = ast_type_to_ir p.Ast.type_ in
    let pv = Ir.fresh_vreg ir_func in
    Ir.(entry.instrs <- Const (pv, 0L, pt) :: entry.instrs);
    Ir.(entry.instrs <- Store (v, pv, pt) :: entry.instrs);
    Hashtbl.add vars p.Ast.name (v, pt)
  ) func.Ast.params;
  (* Build body *)
  ignore (build_block ir_func entry func.Ast.body vars);
  ir_func.Ir.blocks <- [entry];
  ir_func

and build_block ir_func block vars =
  let result_instrs = ref [] in
  let result_term = ref (Ir.Ret 0) in
  List.iter (fun stmt ->
    match stmt with
    | Ast.Return expr ->
      let v, _ = build_expr ir_func block vars expr in
      result_term := Ir.Ret v
    | Ast.VarDecl vd ->
      let v = Ir.fresh_vreg ir_func in
      let t = ast_type_to_ir vd.Ast.type_ in
      block.Ir.instrs <- Ir.Alloc (v, t) :: block.Ir.instrs;
      (match vd.Ast.init with
       | Some e ->
         let ev, _ = build_expr ir_func block vars e in
         block.Ir.instrs <- Ir.Store (v, ev, t) :: block.Ir.instrs
       | None -> ());
      Hashtbl.add vars vd.Ast.name (v, t)
    | Ast.If { cond; then_block; else_branch; _ } ->
      let cv, _ = build_expr ir_func block vars cond in
      let then_label = "b_then" in
      let else_label = "b_else" in
      let end_label = "b_end" in
      block.Ir.terminator <- Ir.BrCond (cv, then_label, else_label);
      let then_b = Ir.make_block then_label in
      ignore (build_block ir_func then_b then_block vars);
      then_b.Ir.terminator <- Ir.Br end_label;
      let else_b = Ir.make_block else_label in
      (match else_branch with
       | Some s ->
         (match s with
          | Ast.Block b -> ignore (build_block ir_func else_b b vars)
          | Ast.If _ ->
            let sub_ir_func = ir_func in
            let sub = Ir.make_block "b_sub" in
            else_b.Ir.terminator <- Ir.Br "b_sub";
            let _ = build_if ir_func sub s vars in
            sub.Ir.terminator <- Ir.Br end_label;
            ir_func.Ir.blocks <- ir_func.Ir.blocks @ [sub]
          | _ -> ())
       | None -> ());
      else_b.Ir.terminator <- Ir.Br end_label;
      let end_b = Ir.make_block end_label in
      ir_func.Ir.blocks <- ir_func.Ir.blocks @ [then_b; else_b; end_b];
      block := end_b
    | Ast.While { cond; body; _ } ->
      let cond_label = "b_cond" in
      let body_label = "b_body" in
      let end_label = "b_end" in
      block.Ir.terminator <- Ir.Br cond_label;
      let cond_b = Ir.make_block cond_label in
      let cv, _ = build_expr ir_func cond_b vars cond in
      cond_b.Ir.terminator <- Ir.BrCond (cv, body_label, end_label);
      let body_b = Ir.make_block body_label in
      let _ = build_block ir_func body_b body vars in
      body_b.Ir.terminator <- Ir.Br cond_label;
      let end_b = Ir.make_block end_label in
      ir_func.Ir.blocks <- ir_func.Ir.blocks @ [cond_b; body_b; end_b];
      block := end_b
    | Ast.Block b -> ignore (build_block ir_func block b vars)
    | Ast.Expr e -> let _, _ = build_expr ir_func block vars e in ()
  ) block.Ast.stmts;
  block.Ir.instrs <- List.rev !result_instrs;
  ()

and build_block ir_func ir_block ast_block vars =
  List.iter (fun stmt ->
    match stmt with
    | Ast.Return expr ->
      let v, _ = build_expr ir_func ir_block vars expr in
      ir_block.Ir.terminator <- Ir.Ret v
    | Ast.VarDecl vd ->
      let v = Ir.fresh_vreg ir_func in
      let t = ast_type_to_ir vd.Ast.type_ in
      ir_block.Ir.instrs <- Ir.Alloc (v, t) :: ir_block.Ir.instrs;
      (match vd.Ast.init with
       | Some e ->
         let ev, _ = build_expr ir_func ir_block vars e in
         ir_block.Ir.instrs <- Ir.Store (v, ev, t) :: ir_block.Ir.instrs
       | None -> ());
      Hashtbl.add vars vd.Ast.name (v, t)
    | Ast.If _ -> build_if ir_func ir_block stmt vars
    | Ast.While _ -> build_while ir_func ir_block stmt vars
    | Ast.Block b -> build_block ir_func ir_block b vars
    | Ast.Expr e ->
      let _, _ = build_expr ir_func ir_block vars e in ()
  ) ast_block.Ast.stmts

and build_if ir_func block stmt vars =
  match stmt with
  | Ast.If { cond; then_block; else_branch; _ } ->
    let cv, _ = build_expr ir_func block vars cond in
    let then_lab = Printf.sprintf "b_then_%d" (Ir.fresh_vreg ir_func) in
    let else_lab = Printf.sprintf "b_else_%d" (Ir.fresh_vreg ir_func) in
    let end_lab = Printf.sprintf "b_end_%d" (Ir.fresh_vreg ir_func) in
    block.Ir.terminator <- Ir.BrCond (cv, then_lab, else_lab);
    let then_b = Ir.make_block then_lab in
    build_block ir_func then_b then_block vars;
    then_b.Ir.terminator <- Ir.Br end_lab;
    ir_func.Ir.blocks <- ir_func.Ir.blocks @ [then_b];
    (match else_branch with
     | Some s ->
       let else_b = Ir.make_block else_lab in
       (match s with
        | Ast.Block b -> build_block ir_func else_b b vars; else_b.Ir.terminator <- Ir.Br end_lab
        | Ast.If _ -> build_if ir_func else_b s vars; else_b.Ir.terminator <- Ir.Br end_lab
        | _ -> ());
       ir_func.Ir.blocks <- ir_func.Ir.blocks @ [else_b]
     | None ->
       let else_b = Ir.make_block else_lab in
       else_b.Ir.terminator <- Ir.Br end_lab;
       ir_func.Ir.blocks <- ir_func.Ir.blocks @ [else_b]);
    let end_b = Ir.make_block end_lab in
    ir_func.Ir.blocks <- ir_func.Ir.blocks @ [end_b]
  | _ -> ()

and build_while ir_func block stmt vars =
  match stmt with
  | Ast.While { cond; body; _ } ->
    let cond_lab = Printf.sprintf "b_cond_%d" (Ir.fresh_vreg ir_func) in
    let body_lab = Printf.sprintf "b_body_%d" (Ir.fresh_vreg ir_func) in
    let end_lab = Printf.sprintf "b_end_%d" (Ir.fresh_vreg ir_func) in
    block.Ir.terminator <- Ir.Br cond_lab;
    let cond_b = Ir.make_block cond_lab in
    let cv, _ = build_expr ir_func cond_b vars cond in
    cond_b.Ir.terminator <- Ir.BrCond (cv, body_lab, end_lab);
    ir_func.Ir.blocks <- ir_func.Ir.blocks @ [cond_b];
    let body_b = Ir.make_block body_lab in
    build_block ir_func body_b body vars;
    body_b.Ir.terminator <- Ir.Br cond_lab;
    ir_func.Ir.blocks <- ir_func.Ir.blocks @ [body_b];
    let end_b = Ir.make_block end_lab in
    ir_func.Ir.blocks <- ir_func.Ir.blocks @ [end_b]
  | _ -> ()

and build_expr ir_func block vars expr =
  match expr with
  | Ast.Int n ->
    let v = Ir.fresh_vreg ir_func in
    block.Ir.instrs <- Ir.Const (v, n, Ir.I32) :: block.Ir.instrs;
    (v, Ir.I32)
  | Ast.Bool b ->
    let v = Ir.fresh_vreg ir_func in
    block.Ir.instrs <- Ir.Const (v, if b then 1L else 0L, Ir.Bool) :: block.Ir.instrs;
    (v, Ir.Bool)
  | Ast.Ident name ->
    (match Hashtbl.find_opt vars name with
     | Some (v, t) ->
       let r = Ir.fresh_vreg ir_func in
       block.Ir.instrs <- Ir.Load (r, v, t) :: block.Ir.instrs;
       (r, t)
     | None ->
       let v = Ir.fresh_vreg ir_func in
       block.Ir.instrs <- Ir.Const (v, 0L, Ir.I32) :: block.Ir.instrs;
       (v, Ir.I32))
  | Ast.Binary (op, l, r) ->
    let lv, lt = build_expr ir_func block vars l in
    let rv, _ = build_expr ir_func block vars r in
    let v = Ir.fresh_vreg ir_func in
    let instr = match op with
      | Ast.Add -> Ir.Add (v, lv, rv) | Ast.Sub -> Ir.Sub (v, lv, rv)
      | Ast.Mul -> Ir.Mul (v, lv, rv) | Ast.Div -> Ir.Div (v, lv, rv)
      | Ast.Eq -> Ir.Eq (v, lv, rv) | Ast.Neq ->
        let tmp = Ir.fresh_vreg ir_func in
        block.Ir.instrs <- Ir.Eq (tmp, lv, rv) :: block.Ir.instrs;
        Ir.Not (v, tmp)
      | Ast.Lt -> Ir.Lt (v, lv, rv) | Ast.Gt ->
        let tmp = Ir.fresh_vreg ir_func in
        block.Ir.instrs <- Ir.Lt (tmp, rv, lv) :: block.Ir.instrs;
        Ir.Const (v, 0L, Ir.I32)
      | Ast.Le ->
        let tmp1 = Ir.fresh_vreg ir_func in
        let tmp2 = Ir.fresh_vreg ir_func in
        block.Ir.instrs <- Ir.Lt (tmp1, rv, lv) :: block.Ir.instrs;
        Ir.Not (v, tmp1)
      | Ast.Ge ->
        let tmp = Ir.fresh_vreg ir_func in
        block.Ir.instrs <- Ir.Lt (tmp, lv, rv) :: block.Ir.instrs;
        Ir.Not (v, tmp)
      | Ast.And -> Ir.And (v, lv, rv)
      | Ast.Or -> Ir.Or (v, lv, rv)
    in
    block.Ir.instrs <- instr :: block.Ir.instrs;
    (v, lt)
  | Ast.Unary (op, e) ->
    let ev, et = build_expr ir_func block vars e in
    let v = Ir.fresh_vreg ir_func in
    let instr = match op with
      | Ast.Neg -> Ir.Neg (v, ev)
      | Ast.Not -> Ir.Not (v, ev)
    in
    block.Ir.instrs <- instr :: block.Ir.instrs;
    (v, et)
  | Ast.Call (name, args) ->
    let arg_vs = List.map (fun a -> fst (build_expr ir_func block vars a)) args in
    let v = Ir.fresh_vreg ir_func in
    block.Ir.instrs <- Ir.Call (v, name, arg_vs, Ir.I32) :: block.Ir.instrs;
    (v, Ir.I32)
  | Ast.Cast (t, e) ->
    let ev, _ = build_expr ir_func block vars e in
    (ev, ast_type_to_ir t)
  | Ast.Assign (l, r) ->
    let rv, rt = build_expr ir_func block vars r in
    (match l with
     | Ast.Ident name ->
       (match Hashtbl.find_opt vars name with
        | Some (v, t) ->
          block.Ir.instrs <- Ir.Store (v, rv, t) :: block.Ir.instrs;
          (rv, rt)
        | None -> (rv, rt))
     | _ -> (rv, rt))

let build_module source prog =
  let ir_funcs = List.filter_map (fun decl ->
    match decl with Ast.Func f -> Some (build_func f)) prog.Ast.decls
  in
  { Ir.funcs = ir_funcs; globals = [] }

File: L1S/v3/code/rust/src/ir_build.rs
use std::collections::HashMap;
use crate::ast::*;
use crate::ir::*;
use crate::typeck::ast_type_to_ir;

pub struct IrBuilder {
    pub module: IrModule,
    func_idx: usize,
    label_counter: usize,
}

impl IrBuilder {
    pub fn new() -> Self {
        Self { module: IrModule { funcs: vec![] }, func_idx: 0, label_counter: 0 }
    }

    pub fn build(&mut self, prog: &AstProgram) {
        for decl in &prog.decls {
            if let AstDecl::Func(func) = decl {
                self.build_func(func);
            }
        }
    }

    fn new_label(&mut self, prefix: &str) -> String {
        self.label_counter += 1;
        format!("{}_{}", prefix, self.label_counter)
    }

    fn func_mut(&mut self) -> &mut IrFunction {
        &mut self.module.funcs[self.func_idx]
    }

    fn emit(&mut self, instr: IrInstr) {
        let func = self.func_mut();
        func.blocks.last_mut().unwrap().instrs.push(instr);
    }

    fn set_term(&mut self, term: IrTerminator) {
        let func = self.func_mut();
        func.blocks.last_mut().unwrap().terminator = Some(term);
    }

    fn start_block(&mut self, label: String) {
        let func = self.func_mut();
        func.blocks.push(IrBlock { label, instrs: vec![], terminator: None });
    }

    fn alloc_vreg(&mut self) -> VReg {
        let func = self.func_mut();
        let vreg = func.num_vregs;
        func.num_vregs += 1;
        vreg
    }

    fn build_func(&mut self, func: &AstFuncDecl) {
        let param_types: Vec<IrType> = func.params.iter().map(|p| ast_type_to_ir(&p.type_)).collect();
        let return_type = ast_type_to_ir(&func.return_type);
        let name = func.name.clone();

        let func_idx = self.module.funcs.len();
        self.module.funcs.push(IrFunction {
            name: name.clone(),
            param_types: param_types.clone(),
            param_vregs: vec![],
            return_type: return_type.clone(),
            blocks: vec![],
            num_vregs: 0,
        });
        self.func_idx = func_idx;

        let entry_label = self.new_label("entry");
        self.start_block(entry_label);

        let mut vars: HashMap<String, VReg> = HashMap::new();
        for param in &func.params {
            let vreg = self.alloc_vreg();
            self.func_mut().param_vregs.push(vreg);
            vars.insert(param.name.clone(), vreg);
        }

        self.build_block(&func.body, &mut vars);

        let needs_term = self.func_mut().blocks.last().unwrap().terminator.is_none();
        if needs_term {
            self.set_term(IrTerminator::Ret(None));
        }
    }

    fn build_block(&mut self, block: &AstBlock, vars: &mut HashMap<String, VReg>) {
        for stmt in &block.stmts {
            self.build_stmt(stmt, vars);
        }
    }

    fn build_stmt(&mut self, stmt: &AstStmt, vars: &mut HashMap<String, VReg>) {
        match stmt {
            AstStmt::Return(expr_opt) => {
                let vreg = expr_opt.as_ref().map(|e| self.build_expr(e, vars));
                self.set_term(IrTerminator::Ret(vreg));
            }
            AstStmt::VarDecl(decl) => {
                let vreg = self.alloc_vreg();
                if let Some(init) = &decl.init {
                    let init_vreg = self.build_expr(init, vars);
                    self.emit(IrInstr::Mov(vreg, init_vreg));
                }
                vars.insert(decl.name.clone(), vreg);
            }
            AstStmt::If(ast_if) => {
                let cond_vreg = self.build_expr(&ast_if.cond, vars);
                let then_label = self.new_label("then");
                let else_label = self.new_label("else");
                let end_label = self.new_label("endif");
                self.set_term(IrTerminator::BrCond(cond_vreg, then_label.clone(), else_label.clone()));

                self.start_block(then_label);
                self.build_block(&ast_if.then_block, vars);
                if self.func_mut().blocks.last().unwrap().terminator.is_none() {
                    self.set_term(IrTerminator::Br(end_label.clone()));
                }

                self.start_block(else_label);
                if let Some(else_branch) = &ast_if.else_branch {
                    self.build_stmt(else_branch, vars);
                }
                if self.func_mut().blocks.last().unwrap().terminator.is_none() {
                    self.set_term(IrTerminator::Br(end_label.clone()));
                }

                self.start_block(end_label);
            }
            AstStmt::While(ast_while) => {
                let loop_label = self.new_label("loop");
                let body_label = self.new_label("while_body");
                let end_label = self.new_label("while_end");

                self.start_block(loop_label.clone());
                let cond_vreg = self.build_expr(&ast_while.cond, vars);
                self.set_term(IrTerminator::BrCond(cond_vreg, body_label.clone(), end_label.clone()));

                self.start_block(body_label);
                self.build_block(&ast_while.body, vars);
                if self.func_mut().blocks.last().unwrap().terminator.is_none() {
                    self.set_term(IrTerminator::Br(loop_label));
                }

                self.start_block(end_label);
            }
            AstStmt::Block(block) => {
                self.build_block(block, vars);
            }
            AstStmt::Expr(expr) => {
                self.build_expr(expr, vars);
            }
        }
    }

    fn build_expr(&mut self, expr: &AstExpr, vars: &HashMap<String, VReg>) -> VReg {
        match expr {
            AstExpr::Int(val) => {
                let vreg = self.alloc_vreg();
                self.emit(IrInstr::Const(vreg, *val, IrType::I32));
                vreg
            }
            AstExpr::Bool(val) => {
                let vreg = self.alloc_vreg();
                self.emit(IrInstr::Const(vreg, if *val { 1 } else { 0 }, IrType::Bool));
                vreg
            }
            AstExpr::Ident(name) => {
                let var_vreg = *vars.get(name).unwrap_or_else(|| {
                    panic!("undefined variable: {}", name)
                });
                let vreg = self.alloc_vreg();
                self.emit(IrInstr::Mov(vreg, var_vreg));
                vreg
            }
            AstExpr::Binary(op, lhs, rhs) => {
                let lv = self.build_expr(lhs, vars);
                let rv = self.build_expr(rhs, vars);
                let result = self.alloc_vreg();
                let instr = match op {
                    BinaryOp::Add => IrInstr::Add(result, lv, rv),
                    BinaryOp::Sub => IrInstr::Sub(result, lv, rv),
                    BinaryOp::Mul => IrInstr::Mul(result, lv, rv),
                    BinaryOp::Div => IrInstr::Div(result, lv, rv),
                    BinaryOp::Eq => IrInstr::Eq(result, lv, rv),
                    BinaryOp::Neq => {
                        let eq_vreg = self.alloc_vreg();
                        self.emit(IrInstr::Eq(eq_vreg, lv, rv));
                        self.emit(IrInstr::Not(result, eq_vreg));
                        return result;
                    }
                    BinaryOp::Lt => IrInstr::Lt(result, lv, rv),
                    BinaryOp::Gt => {
                        let lt_vreg = self.alloc_vreg();
                        self.emit(IrInstr::Lt(lt_vreg, rv, lv));
                        return lt_vreg;
                    }
                    BinaryOp::Le => {
                        let gt_vreg = self.alloc_vreg();
                        self.emit(IrInstr::Lt(gt_vreg, rv, lv));
                        self.emit(IrInstr::Not(result, gt_vreg));
                        return result;
                    }
                    BinaryOp::Ge => {
                        let lt_vreg = self.alloc_vreg();
                        self.emit(IrInstr::Lt(lt_vreg, lv, rv));
                        self.emit(IrInstr::Not(result, lt_vreg));
                        return result;
                    }
                    BinaryOp::And => IrInstr::And(result, lv, rv),
                    BinaryOp::Or => IrInstr::Or(result, lv, rv),
                };
                self.emit(instr);
                result
            }
            AstExpr::Unary(op, operand) => {
                let ov = self.build_expr(operand, vars);
                let result = self.alloc_vreg();
                match op {
                    UnaryOp::Neg => self.emit(IrInstr::Neg(result, ov)),
                    UnaryOp::Not => self.emit(IrInstr::Not(result, ov)),
                }
                result
            }
            AstExpr::Call(name, args) => {
                let arg_vregs: Vec<VReg> = args.iter().map(|a| self.build_expr(a, vars)).collect();
                let result = self.alloc_vreg();
                let ret_type = IrType::I32;
                self.emit(IrInstr::Call(result, name.clone(), arg_vregs, ret_type));
                result
            }
            AstExpr::Cast(_, expr) => {
                self.build_expr(expr, vars)
            }
            AstExpr::Assign(lhs, rhs) => {
                let rhs_vreg = self.build_expr(rhs, vars);
                if let AstExpr::Ident(name) = lhs.as_ref() {
                    let var_vreg = *vars.get(name).unwrap_or_else(|| {
                        panic!("undefined variable: {}", name)
                    });
                    self.emit(IrInstr::Mov(var_vreg, rhs_vreg));
                    var_vreg
                } else {
                    panic!("invalid assignment target");
                }
            }
        }
    }
}
